### DISCUSSION TITLE
MySql 5.7+ and date/datetime dates (again)

### DISCUSSION THREAD

**Discussion by lat9** [zencart/zencart#5837] [Ideas]
Just reading through the MySQL 5.7 documentation on date/datetime fields (https://dev.mysql.com/doc/refman/5.7/en/datetime.html) and I think we might still have some issues.  From that documentation:
```
 The DATE type is used for values with a date part but no time part. MySQL retrieves and displays DATE values in 'YYYY-MM-DD' format. 
The supported range is '1000-01-01' to '9999-12-31'.

The DATETIME type is used for values that contain both date and time parts. MySQL retrieves and displays DATETIME values in 'YYYY-MM-DD hh:mm:ss' format. 
The supported range is '1000-01-01 00:00:00' to '9999-12-31 23:59:59'. 
```
There are still (as of zc156b, anyway) a slew of date-related fields that use a default date of `0001-01-01`, which will (from the documentation) cause issues.

**Comment by Zen4All** [comment]
I have been using 5.7 for years now, and it never gave any issues with 0001-01-01

**Comment by lat9** [comment]
That's great, @Zen4All, I'm just pointing to the documentation that says differently.

**Comment by Zen4All** [comment]
I read the link above to, and did some searching on internet, but could not find any good references to 0001-01-01. I am not sure why we used 0001-01-01, but I think I copied it form the zen cart code for my mods.
Maybe someone made a typo in the past?

**Comment by drbyte** [comment]
We inherited using `0000-00-00` from osc.
When MySQL started forcing strict no-zero-dates rules, they still had not yet documented an acceptable alternative to zero-dates, and we did a lot of testing to be sure that `0001-01-01` would work without negative side-effects.
Now it seems MySQL has gone back and defined what they wish to be seen as acceptable values. It's unfortunate that they chose something different, particularly when nobody's going to use either of those choices for "real" data anyway.

Anyway ...

The better approach would be to use `null` instead of 0000-00-00 or 0001-01-01 (or 1000-01-01 IMO), because the whole point of these "old" dates is to indicate that "nothing has been supplied". (And an empty string is the wrong "type".)
Unfortunately converting all the code to handle null dates correctly is also a big undertaking, and affects addons since many mods deal with data concerning dates.

I'd welcome a series of PRs that handle implementing null dates, if the PRs are done feature-by-feature (ie: sale dates in one, coupon dates in another, birthdates in another, specials, then featured products, etc). This will help plugin authors see what's changed related to their code.

**Comment by drbyte** [comment]
Aside: I hope we don't have to figure out how to accommodate `23:59:60` as well!
https://alexwlchan.net/2019/05/falsehoods-programmers-believe-about-unix-time/

**Comment by torvista** [comment]
I got an installer error:

> [25-Apr-2020 16:37:12 UTC] MySQL error 1292 encountered during zc_install:
> Incorrect datetime value: '0000-00-00 00:00:00' for column 'gdpr_deleted_date' at row 1
> ALTER TABLE customers ADD customers_secret varchar(64) NOT NULL default '';

Evidently the problem field is one of mine, unrelated to the query in question, but nonetheless it stopped the upgrade query on that table.

It's a bit messy to fix and may even not come to the attention to the upgrader immediately, does this not indicate that it would be a good idea if the installer should check for zero dates and zero date as default value, as part of the initial inspection?

**Comment by scottcwilson** [comment]
It does check for fields which are part of the default Zen Cart installation.  This field was created by a mod.

**Comment by torvista** [comment]
I know its a mod...when I say perhaps the installer should check for this, I mean prior to an upgrade.
I have not looked but I imagine every release adds more checks to filter out things like this that become issues as MySQl and php get more strict.

**Comment by lat9** [comment]
> 
> I know its a mod...when I say perhaps the installer should check for this, I mean prior to an upgrade.
> I have not looked but I imagine every release adds more checks to filter out things like this that become issues as MySQl and php get more strict.

Please note that it's one thing to _inspect_ the database for all tables for these zero-date conditions -- issuing a warning to the person updating -- it's quite another to willy-nilly _assume_ that you're correcting the condition by changing those `0000-00-00` dates to something that's now valid.  The code associated with the database field might be specifically looking for that date-value and changing entries' values could make the problem even worse.

**Comment by torvista** [comment]
I only meant check/inspect to flag it up, of course absolutely not do any modifications.

**Comment by drbyte** [comment]
MySQL is evidently conflicted on the best solution: In their documentation they've said the NO_ZERO_DATES setting is deprecated and if you set it you'll get errors, but it's also the default, and they can't decide how/when to add a new setting to do whatever they're thinking of doing next with it.
... all the way from 5.6 to 8.0 ...
https://dev.mysql.com/doc/refman/8.0/en/sql-mode.html#sqlmode_no_zero_date

**Comment by scottcwilson** [comment]
Date handling for plugins is discussed here: 

https://docs.zen-cart.com/user/upgrading/date_standardization/
